Add the in-memory metrics exporter.

yxue opened the pull request:
Fixes # (issue) [1405](https://github.com/open-telemetry/opentelemetry-cpp/issues/1405)

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

owent commented:
Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .

yxue commented:
> Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .

Sure, I can add that. I feel supporting/maintaining two build systems is a little bit inconvenient. Do we consider using some tools like, https://github.com/google/bazel-to-cmake, so that CMakeLists.txt can be generated from the BUILD and WORKSPACE, or using cmake and bazel (https://github.com/bazelbuild/rules_foreign_cc) supports importing cmake projects.

owent commented:
> > Could you please alse add cmake targets in `exporters/memory/CMakeLists.txt` and `cmake/opentelemetry-cpp-config.cmake.in` .
> 
> Sure, I can add that. I feel supporting/maintaining two build systems is a little bit inconvenient. Do we consider using some tools like, https://github.com/google/bazel-to-cmake, so that CMakeLists.txt can be generated from the BUILD and WORKSPACE, or using cmake and bazel (https://github.com/bazelbuild/rules_foreign_cc) supports importing cmake projects. 

Good point, I think it require more discussion. In most case, I think cmake users do not want to install another bazel toolset, and bazel users do not want install cmake toolset either.
cmake has many modules and tools to work with legacy build toolchains(pkg-config, makefile, autoconf, android ndk, feature dection by test running or compiling some c/c++ codes, environments and etc.).I'm not family with bazel and I'm not sure if it's easy to let bazel and cmake work together with these modules.

BTW: We had a discussion before and try to find a way to use bazel to build both with legacy gcc 4.8 and the latest compiler here https://github.com/open-telemetry/opentelemetry-cpp/pull/666 , do you has any suggestion about it?

lalitb commented:
>  I feel supporting/maintaining two build systems is a little bit inconvenient

Yes, I do agree it's inconvenient to maintain both these build systems. The decision to support both was made much before any of the existing contributors joined the team. The SIG work was initiated by the folks from Google and Microsoft, and it was defacto choice to support the build systems used by these companies :). Having said that, most of the existing C++ projects use either(or both) of these build systems, and so it makes sense to keep supporting both the build systems.

ThomsonTan commented:
The 2 build systems issue were discussed in the past as @lalitb and @owent mentioned. As there are users for both of them, we will try our best to keep both up to date and consistent. For the https://github.com/google/bazel-to-cmake, it has not been updated for a few years, not sure it could work, together with that we've already have full set of cmake files which is not expected to be overwriten.

yxue commented:
> 



> Thanks for the PR @yxue. This is your first PR here, and it is nicely done :)
> 
> Just thinking whether the InMemoryExporter should have any dependency on the protobuf. For Logs and Traces, the memory/ostream exporters are not dependent on any external library. Is it possible to achieve this for Metrics too - Eg by letting CircularBuffer contain the list of [MetricData ](https://github.com/open-telemetry/opentelemetry-cpp/blob/8691a2d582c2e0c811fbf383b572e1ff882b5b37/sdk/include/opentelemetry/sdk/metrics/data/metric_data.h#L30) or [ResourceMetrics ](https://github.com/open-telemetry/opentelemetry-cpp/blob/8691a2d582c2e0c811fbf383b572e1ff882b5b37/sdk/include/opentelemetry/sdk/metrics/export/metric_producer.h#L27) ?
> 
> We can discuss further if it is something not feasible/or difficult to achieve with the current SDK implementation.

Thanks @lalitb! I was thinking about using the CircularBuffer with the  ResourceMetrics but then we need the special process for the pointers inside the structure.  

Having the MetricData with CircularBuffer might work but some info is lost. 

Another thing is about comparing these structure objects. With protobuf message, the objects can be easily compared. Since we don't define the == for MetricsData and ResourceMetrics, comparing them in the unit test is not trivial.

yxue commented:
Thanks @lalitb @owent and @ThomsonTan for the info about the building system. 

I am not sure about fixing the gcc 4.8 and grpc compatibility. I do see the building systems are right now used by multiple teams/companies. I also saw there are already some divergence problems between these two systems, e.g., https://github.com/open-telemetry/opentelemetry-cpp/issues/1428. I think we might need some tools to reconcile that if the community is going to keep two building systems.

lalitb commented:
> Thanks @lalitb! I was thinking about using the CircularBuffer with the ResourceMetrics but then we need the special process for the pointers inside the structure

Yes, I thought so too. Should be fine to keep the dependency unless someone has any concern :)
btw, I didn't realize this PR also has changes for bazel build for metrics proto files and raised a PR for that in #1489. Sorry about that, as you would have to resolve the conflict here :( Would ensure not to merge any further conflicting PRs till this is merged.

yxue commented:
> > Thanks @lalitb! I was thinking about using the CircularBuffer with the ResourceMetrics but then we need the special process for the pointers inside the structure
> 
> Yes, I thought so too. Should be fine to keep the dependency unless someone has any concern :) btw, I didn't realize this PR also has changes for bazel build for metrics proto files and raised a PR for that in #1489. Sorry about that, as you would have to resolve the conflict here :( Would ensure not to merge any further conflicting PRs till this is merged.

Sure, no worries. The PR was updated. PTAL, thanks!

lalitb commented on exporters/memory/test/in_memory_metric_exporter_test.cc:
```diff
@@ -0,0 +1,298 @@
+
+#  include <gmock/gmock.h>
+#  include <gtest/gtest.h>
+#  include <iostream>
```
do we need iostream?

lalitb reviewed (approved):
LGTM.

yxue commented on exporters/memory/src/in_memory_metric_exporter.cc:
```diff
@@ -0,0 +1,55 @@
+
+bool InMemoryMetricExporter::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  const std::lock_guard<opentelemetry::common::SpinLockMutex> locked(lock_);
```
done

yxue commented on exporters/memory/include/opentelemetry/exporters/memory/in_memory_data.h:
```diff
@@ -0,0 +1,62 @@
+/**
+ * A wrapper class holding in memory exporter data
+ */
+template <typename T>
```
sure, https://github.com/open-telemetry/opentelemetry-cpp/issues/1493 created

owent commented:
Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.

yxue commented:
> Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.

The change under exporters/otlp is for fixing the missing part of conversion between protobuf and structure. Remove that will fail the unit test.

owent commented:
> > Could you please revert changes in `/exporters/otlp` ?It's duplicated with #1487 , I belive this PR is just for memory metrics exporter.
> 
> The change under exporters/otlp is for fixing the missing part of conversion between protobuf and structure. Remove that will fail the unit test.

OK. we can solve this conflict later.

owent reviewed (changes requested):
LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in` ?

lalitb commented:
> LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in`.

Good point. @yxue this file contains the rules to enable opentelemetry-cpp included in other projects cmake files using `find_package(opentelemetry-cpp)`. All the targets/libraries created by opentelemetry-cpp should be included in this file.

yxue commented:
> > LGTM after `in_memory_metric_exporter` is also added into `cmake/opentelemetry-cpp-config.cmake.in`.
> 
> Good point. @yxue this file contains the rules to enable opentelemetry-cpp included in other projects cmake files using `find_package(opentelemetry-cpp)`. All the targets/libraries created by opentelemetry-cpp should be included in this file.

Updated, thanks!

owent commented on exporters/memory/CMakeLists.txt:
```diff
@@ -21,19 +21,56 @@ install(
+    opentelemetry_exporter_in_memory_metric
+    PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>"
+           "$<INSTALL_INTERFACE:include>"
+           "$<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/../otlp/include>")
```
`target_link_libraries(opentelemetry_exporter_in_memory_metric PUBLIC opentelemetry_otlp_recordable)` would be a better alternative here.But when we building without `WITH_OTLP`, `opentelemetry_otlp_recordable` will be unavailable.
Maybe this module should also be disabled when we build without `WITH_OTLP` now?

marcalff commented on exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h:
```diff
@@ -0,0 +1,67 @@
+  bool Shutdown(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+
+  /**
+   * @return Returns a shared pointer to this exporters InMemoryMetricData.
+   */
+  inline InMemoryMetricData &GetData() noexcept { return *data_; }
```
The comment is outdated, this does not return a shared pointer.

marcalff reviewed (commented):
LGTM

ThomsonTan commented:
@yxue could you please resolve the merge conflicts?

marcalff reviewed (changes requested):
As discussed in C++ SIG meeting,
the in memory storage structure should not depend on protobuf:

```
  using InMemoryMetricData =
      opentelemetry::exporter::memory::InMemoryData<proto::metrics::v1::ResourceMetrics>;

```
Some refactoring needed to address this part.
